doc/ir-semantics.text

```diff
@@ -121,49 +121,103 @@ declares i =
 requires i =
   | mut x = e
   | const x = e
-  | osr e, _        -> vars(e)
+  | osr (e, v, l, (x = ee)*)
+                    -> vars(e) ∪ vars(ee)*
   | x <- e          -> x :: vars(e)
   | branch x _
   | print x         -> [x]
   | _               -> []
 
-update S i = S \ declares i ∪ requires i
-
-## Well-scoped module
-
-M, S ⊨(scp) l
--------------   ∀ l ∈ M : ∃ S
-     ⊨(scp) P
-
-M, S_l* ⊨(scp) l*
------------------   ∀ l* ∈ pred (M, l) : ∃ S_l* ⊂ update S
-   M, S ⊨(scp) l
-
-
--------------
-M, ∅ ⊨(scp) s
-
-
-## Frame construction
-
-⊨(fra) M    P ⊨(osr) M
-----------------------   ∀ M ∈ P
-       ⊨(fra) P
-
-
-M, S ⊨(fra) l
--------------   ∀ l ∈ M : ∃ S
-     ⊨(fra) P
-
-M, update S ⊨(fra) l*
----------------------   ∀ l* ∈ pred (M, l)
-       M, S ⊨(fra) l
-
-
--------------
-M, ∅ ⊨(fra) s
-
-
-P(v), x̄ ⊨(fra) l
-----------------   ∀ osr(e, v, l, (x = ee)*) ∈ M
-      P ⊨(osr) M
+scope S ::= x*
+scope assignment A ::= ((v, l) → S)*
+
+## Declarative scoping judgment
+
+This judgment just classifies "possible" scope assignments A for
+a program P. It does not characterize the tightest possible
+assignment¹, as it allows to implicitly drop variables from scope on
+each transition (the ⊇ relations in the (v, l → i) rule). In
+particular, it cannot forbid shadowing, as it is always possible to
```

Comment:
why can't it? let's just require:

    ∀l* ∈ pred(P(v),l), A(v,l*)  <intersection> (declares i) is empty